### PR TITLE
fix: add support for idle connection monitoring (opt-in for now)

### PR DESCRIPTION
**Pull request by ianbotsf**
## Issue \#

Resolves https://github.com/awslabs/aws-sdk-kotlin/issues/1214

## Description of changes

This PR adds a new opt-in idle connection monitor for OkHttp, configurable by the engine parameter `connectionIdlePollingInterval`. This setting is opt-in for now to preserve existing behavior but will be switched to opt-out (or replaced) in a future minor release.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            .tag(SdkRequestTag::class.java)
+            ?.callContext
+            ?: Dispatchers.IO
+
+    override fun connectionAcquired(connection: Connection, call: Call) {
+        // Non-locking map access is okay here because this code will only execute synchronously as part of a
+        // `connectionAcquired` event and will be complete before any future `connectionReleased` event could fire for
+        // the same connection.
+        monitors.remove(connection)?.let { monitor ->
+            5.seconds
```
Comment (lauzadis) [inline runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt:47]: nit: unused value

File: runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            val context = call.callContext()
+            val logger = context.logger<ConnectionIdleMonitor>()
+            logger.trace { "Cancel monitoring for $connection" }
+
+            // Use `runBlocking` because this _must_ finish before OkHttp goes to use the connection
+            val cancelTime = measureTime {
+                runBlocking(context) { monitor.cancelAndJoin() }
+            }
```
Comment (lauzadis) [inline runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt:55]: Why is measuring the cancel time important here?

File: runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        var resetTimeout = true
+        val oldTimeout = socket.soTimeout
+
+        try {
+            socket.soTimeout = pollInterval.inWholeMilliseconds.toInt()
+
+            while (coroutineContext.isActive) {
+                try {
+                    logger.trace { "Polling socket for $conn" }
+                    source.readByte() // Blocking read; will take up to READ_TIMEOUT_MS to complete
```
Comment (lauzadis) [inline runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt:97]: > will take up to READ_TIMEOUT_MS to complete

question/clarification request: I don't see a `READ_TIMEOUT_MS` / any other timeout used here in the `readByte` call

File: runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            }
+
+            logger.trace { "Monitoring coroutine has been cancelled. Ending polling loop." }
+        } catch (e: Throwable) {
+            logger.warn(e) { "Failed to poll $conn. Ending polling loop. Connection may be unstable now." }
+        } finally {
+            if (resetTimeout) {
+                logger.trace { "Attempting to reset soTimeout..." }
+                try {
+                    conn.socket().soTimeout = oldTimeout
```
Comment (lauzadis) [inline runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt:116]: question: Is it possible for a monitored connection to be pulled and used by OkHttp before we reset the `soTimeout`?

**Review by lauzadis — APPROVED** [review]

File: runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt (hunk 4 above)
Comment (0marperez) [reply runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt:114]: Do we have measurements for how long those health checks last and how long we take to reset the `soTimeout`? We might have a lot of overhead or maybe not.

Comment (0marperez) [reply runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt:114]: Nvm this comment, I was just misunderstanding something

File: runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        try {
+            socket.soTimeout = pollInterval.inWholeMilliseconds.toInt()
+
+            while (coroutineContext.isActive) {
+                try {
+                    logger.trace { "Polling socket for $conn" }
+                    source.readByte() // Blocking read; will take up to `pollInterval` time to complete
+                } catch (_: SocketTimeoutException) {
+                    logger.trace { "Socket still alive for $conn" }
+                    // Socket is still alive
```
Comment (0marperez) [inline runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt:98]: Nit: This comment seems unnecessary. Feel free to ignore though.

**Review by 0marperez — APPROVED** [review]

**Pull request merged**